[CI] SearchHitsTests testConcurrentEquals failing

**Build Scans:**
- [elasticsearch-intake #38869 / part1](https://gradle-enterprise.elastic.co/s/vswpfitdpss2o)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.search.SearchHitsTests.testConcurrentEquals" -Dtests.seed=8BC93B1084FF08BD -Dtests.locale=it-IT -Dtests.timezone=Pacific/Noumea -Druntime.java=26
```

**Applicable branches:**
8.19

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.search.SearchHitsTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testConcurrentEquals'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: an empty collection
     but: <[LEAK: resource was not cleaned up before it was garbage-collected.
Recent access records: 
Created at:
	in [TEST-SearchHitsTests.testFromXContent-seed#[8BC93B1084FF08BD]][testFromXContent]
	org.elasticsearch.search.SearchHits.<init>(SearchHits.java:75)
	org.elasticsearch.search.SearchHitsTests.createTestItem(SearchHitsTests.java:80)
	org.elasticsearch.search.SearchHitsTests.createTestItem(SearchHitsTests.java:59)
	org.elasticsearch.search.SearchHitsTests.createXContextTestInstance(SearchHitsTests.java:220)
	org.elasticsearch.search.SearchHitsTests.createXContextTestInstance(SearchHitsTests.java:34)
	org.elasticsearch.test.AbstractXContentTestCase$XContentTester.lambda$test$4(AbstractXContentTestCase.java:159)
	org.elasticsearch.test.ESTestCase.randomValueOtherThanMany(ESTestCase.java:1517)
	org.elasticsearch.test.AbstractXContentTestCase$XContentTester.test(AbstractXContentTestCase.java:151)
	org.elasticsearch.test.Ab
[truncated]
```

**Issue Reasons:**
- [8.19] 1 consecutive failures in test testConcurrentEquals
- [8.19] 1 consecutive failures in step part1
- [8.19] 1 consecutive failures in pipeline elasticsearch-intake

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/test/framework/src/main/java/org/elasticsearch/test/AbstractXContentTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/AbstractXContentTestCase.java
@@ -148,15 +148,22 @@ public abstract class AbstractXContentTestCase<T extends ToXContent> extends EST
                 T testInstance = null;
                 try {
                     if (xContentType.equals(XContentType.YAML)) {
-                        testInstance = randomValueOtherThanMany(instance -> {
-                            // unicode character U+0085 (NEXT LINE (NEL)) doesn't survive YAML round trip tests (see #97716)
-                            // get a new random instance if we detect this character in the xContent output
+                        // unicode character U+0085 (NEXT LINE (NEL)) doesn't survive YAML round trip tests (see #97716)
+                        // Loop until we find an instance whose serialized form doesn't contain this character,
+                        // disposing any rejected candidates so that ref-counted instances are not leaked.
+                        boolean containsNel = true;
+                        while (containsNel) {
+                            if (testInstance != null) {
+                                dispose.accept(testInstance);
+                            }
+                            testInstance = instanceSupplier.apply(xContentType);
                             try {
-                                return toXContent.apply(instance, xContentType).utf8ToString().contains("\u0085");
+                                containsNel = toXContent.apply(testInstance, xContentType).utf8ToString().contains("\u0085");
                             } catch (IOException e) {
+                                // testInstance is non-null, so the outer finally will dispose it
                                 throw new AssertionError(e);
                             }
-                        }, () -> instanceSupplier.apply(xContentType));
+                        }
                     } else {
                         testInstance = instanceSupplier.apply(xContentType);
                     }

